src/views/darkroom.c

```diff
@@ -203,17 +203,13 @@ static dt_darkroom_layout_t _lib_darkroom_get_layout(dt_view_t *self)
     return DT_DARKROOM_LAYOUT_EDITING;
 }
 
-static cairo_filter_t _get_filtering_level(dt_develop_t *dev)
+static cairo_filter_t _get_filtering_level(int closeup)
 {
-  const dt_dev_zoom_t zoom = dt_control_get_dev_zoom();
-  const int closeup = dt_control_get_dev_closeup();
-  const float scale = dt_dev_get_zoom_scale(dev, zoom, 1<<closeup, 0);
-
   // for pixel representation above 1:1, that is when a single pixel on the image
   // is represented on screen by multiple pixels we want to disable any cairo filter
   // which could only blur or smooth the output.
 
-  if(scale / darktable.gui->ppd > 1.0)
+  if(closeup)
```

Comment:
Isn't closeup 0 at 100% and 1 when zoom is 200% on low-dpi? If so, this seems to contradict the comment on this PR.

We must be very careful here as we had some changes recently (3.4?) as the image was somewhat blurred due to fast filtering used at some zoom level.

Reply:
> Isn't closeup 0 at 100% and 1 when zoom is 200% on low-dpi?

Yes. FWIW, this seems to be the "canonical" code to handle closeup:

https://github.com/darktable-org/darktable/blob/52f261bdba7e50c6075169964c5701f74bd9c938/src/views/darkroom.c#L3612
https://github.com/darktable-org/darktable/blob/52f261bdba7e50c6075169964c5701f74bd9c938/src/views/darkroom.c#L3674-L3698
https://github.com/darktable-org/darktable/blob/52f261bdba7e50c6075169964c5701f74bd9c938/src/views/darkroom.c#L3703

So closeup > 0 at >= 200% (lodpi) or >= 400% (hidpi).

> If so, this seems to contradict the comment on this PR.

Yes! And the content of this PR isn't right either, alas.

> We must be very careful here as we had some changes recently (3.4?) as the image was somewhat blurred due to fast filtering used at some zoom level.

Yes indeed. Thank you for catching.

So the current (e.g. v3.6) code is doing the right thing for lodpi. It turns on `CAIRO_FILTER_FAST` at > 100%, as described in `_get_filtering_level()`:

https://github.com/darktable-org/darktable/blob/52f261bdba7e50c6075169964c5701f74bd9c938/src/views/darkroom.c#L212-L214

This comment in `scrolled()` is wrong though:

https://github.com/darktable-org/darktable/blob/52f261bdba7e50c6075169964c5701f74bd9c938/src/views/darkroom.c#L3674

The v3.6 code does not do the right thing for hidpi, so there does need to be a fix. It turns on `CAIRO_FAST_FILTER` at >= 800%. But if 200% is 1:1 at lodpi, `CAIRO_FAST_FILTER` should be on at >= 200%. Cairo catches that 200% is 1:1 and therefore flips to nearest neighbor mode anyhow. But at 400% it bogs down.

So the simple thing would be just to enable `CAIRO_FAST_FILTER` when `scale` is >= 1.

Here is a rundown of v3.6 behavior:

```
lodpi, ppd=1

 50% scale: 0.5  scale/ppd 0.5    closeup 0
                 ------------<---- should fast filter at >= 100%, scale >= 1
100% scale: 1    scale/ppd 1      closeup 0
150% scale: 1.5  scale/ppd 1.5    closeup 0
200% scale: 2    scale/ppd 2      closeup 1
400% scale: 4    scale/ppd 4      closeup 2
800% scale: 8    scale/ppd 8      closeup 3

closeup set at >= 200%
fallback to fast filter at > 100%
```

```
hidpi, ppd=2

100% scale: 0.5  scale/ppd 0.25   closeup 0   scale*ppd 1    
150% scale: 0.75 scale/ppd 0.375  closeup 0   scale*ppd 1.5
                 ------------<---- should fast filter at >= 200%, scale >= 1
200% scale: 1    scale/ppd 0.5    closeup 0   scale*ppd 2
400% scale: 2    scale/ppd 1      closeup 1   scale*ppd 4
                 ------------<---- v3.6 switches to fast filter
800% scale: 4    scale/ppd 2      closeup 2   scale*ppd 8

closeup set at >= 400%
fallback to fast filter at > 400% (e.g. 800%+)
```

There still is a bug in v3.6 where the second window filtering depends on the first window's scale.

Then the other question is whether `closeup` actually represents what it says.

I'll try to sort out the right fix for hidpi. I'm assuming that we _do_ want to switch to `CAIRO_FAST_FILTER` at any zoom level >= 1:1. There could also be an argument that beween 1:1 and 1:2 (e.g. 100% < zoom < 200% for lodpi, but a moot point for hidpi which doesn't have steps between 200% and 400%) that we want actually do want to use as nicer interpolation. I'm also assuming that 200% == 1:1 at hidpi. If not this all gets funny again.